game/CMakeLists.txt

```diff
@@ -60,17 +60,52 @@ endif()
 
 set(SOURCES ${SOURCE_FILES} ${HEADER_FILES} ${RESOURCE_FILES})
 
+# Define main executable
+add_executable(performous ${SOURCES})
+
+  string(ASCII 27 Esc)
+# Enable C++14 support and other compiler flags...
+if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+	set(CMAKE_CXX_FLAGS "-fcx-limited-range ${CMAKE_CXX_FLAGS}")
+	execute_process(COMMAND ${CMAKE_CXX_COMPILER} -dumpversion OUTPUT_VARIABLE GCC_VERSION)
+	message("Detected ${Esc}[36mGCC ${Esc}[mversion ${Esc}[36m${GCC_VERSION}${Esc}[m")
+	if (GCC_VERSION VERSION_GREATER_EQUAL 8.0)
+		message(STATUS "Setting C++ standard to: ${Esc}[36mC++17${Esc}[m")
+		set_property(TARGET performous PROPERTY CXX_STANDARD 17)
+	elseif (GCC_VERSION VERSION_GREATER_EQUAL 7.0)
+		message(STATUS "Setting C++ standard to: ${Esc}[36mC++14${Esc}[m")
+		set_property(TARGET performous PROPERTY CXX_STANDARD 14)
+	elseif (GCC_VERSION VERSION_GREATER_EQUAL 4.8)
+		set_property(TARGET performous PROPERTY CXX_STANDARD 11)
+		set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-attributes -Wno-deprecated-declarations")
+		message(WARNING "Your GCC is rather old, disabling some useful compiler warnings (GCC 5+ or Clang 3.4+ recommended for proper C++14 support).")
+	else()
+		message(FATAL_ERROR "Your C++ compiler is too old, please upgrade.")
+	endif()
+endif()
+
+if(CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+	message("Detected ${Esc}[36mClang/AppleClang ${Esc}[mversion ${Esc}[36m${CMAKE_CXX_COMPILER_VERSION}${Esc}[m")
+	if(CMAKE_CXX_COMPILER_ID STREQUAL "AppleClang" OR CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 5.0)
+		message(STATUS "Setting C++ standard to: ${Esc}[36mC++17${Esc}[m")
+		set_property(TARGET performous PROPERTY CXX_STANDARD 17)
+	elseif (CMAKE_CXX_COMPILER_ID STREQUAL "Clang" AND CMAKE_CXX_COMPILER_VERSION VERSION_GREATER_EQUAL 3.4 )
+		message(STATUS "Setting C++ standard to: C${Esc}[36m++14${Esc}[m")
+		set_property(TARGET performous PROPERTY CXX_STANDARD 14)
+	else ()
+	# FIXME: This case is probably not correct but seriously, I don't think it will come up very frequently anyway.
+		set_property(TARGET performous PROPERTY CXX_STANDARD 11)
+		set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-attributes -Wno-deprecated-declarations")
+		message(WARNING "Your GCC is rather old, disabling some useful compiler warnings (GCC 5+ or Clang 3.4+ recommended for proper C++14 support).")
+	endif()
+endif()
```

Comment:
Can we also add a msvc section?

Reply:
Can you make it?

Reply:
Hmmm currently not on windows @allcreator any experience with this?

Reply:
Hmmm currently not on windows @allcreater any experience with this?

Reply:
> Hmmm currently not on windows @allcreater any experience with this?

https://docs.microsoft.com/en-us/cpp/overview/visual-cpp-language-conformance?view=vs-2019

This should let you do it?

Reply:
> 
> 
> Hmmm currently not on windows @allcreater any experience with this?

I could think about it :)
But in theory I think it's better to minimize platrofm-specific declarations instead of multiplying them :) As far as I remember there is cross-platforms hints for language standard etc.

Reply:
Well, I guess the ideal thing would be to declare which language features we actually are using and let CMake figure the standard out.

Reply:
This is what we need to check but frankly I'm not entirely sure about all the ones we're using:

https://cmake.org/cmake/help/v3.1/prop_gbl/CMAKE_CXX_KNOWN_FEATURES.html#prop_gbl:CMAKE_CXX_KNOWN_FEATURES

Reply:
And here is a quickly done, preliminary list: Feel free to add/remove stuff,

```
cxx_attributes
cxx_auto_type
cxx_constexpr
cxx_defaulted_functions
cxx_deleted_functions
cxx_enum_forward_declarations
cxx_explicit_conversions
cxx_extended_friend_declarations
cxx_lambdas
cxx_lambda_init_captures
cxx_noexcept
cxx_nonstatic_member_init
cxx_nullptr
cxx_range_for
cxx_right_angle_brackets
```

Reply:
Can somebody else weigh in re: this? @performous/core-owners